File: chart/src/main/java/cn/jingzhuan/lib/chart/data/TreeDataSet.java
package cn.jingzhuan.lib.chart.data;

import android.graphics.Color;

import java.util.ArrayList;
import java.util.List;

import cn.jingzhuan.lib.chart.Viewport;
import cn.jingzhuan.lib.chart.component.AxisY;
import cn.jingzhuan.lib.chart.component.AxisY.AxisDependency;


/**
 * Statistic Data Set
 * Created by guobosheng on 22/6/23.
 */

public class TreeDataSet extends AbstractDataSet<TreeValue> {

    private List<TreeValue> mTreeValues;
    private float strokeThickness = 2;
    private int positiveColor = Color.RED;
    private int negativeColor = Color.GREEN;
    private int colorAlpha = 255;

    public TreeDataSet(List<TreeValue> treeValues) {
        this(treeValues, AxisY.DEPENDENCY_BOTH);
    }

    public TreeDataSet(List<TreeValue> treeValues, @AxisDependency int axisDependency) {
        this.mTreeValues = treeValues;
        setAxisDependency(axisDependency);
    }

    @Override
    public int getEntryCount() {
        if (mTreeValues != null) {
            if (getMinValueCount() > mTreeValues.size()) {
                return getMinValueCount();
            } else {
                return mTreeValues.size();
            }
        }
        return 0;
    }

    @Override
    public void calcMinMax(Viewport viewport) {

        if (mTreeValues == null || mTreeValues.isEmpty())
            return;

        mViewportYMax = -Float.MAX_VALUE;
        mViewportYMin = Float.MAX_VALUE;

        for (TreeValue e : getVisiblePoints(viewport)) {
            calcMinMaxY(e);
        }
        float range = mViewportYMax - mViewportYMin;
        if (Float.compare(getMinValueOffsetPercent(), 0f) > 0f) {
            mViewportYMin = mViewportYMin - range * getMinValueOffsetPercent();
        }
        if (Float.compare(getMaxValueOffsetPercent(), 0f) > 0f) {
            mViewportYMax = mViewportYMax + range * getMaxValueOffsetPercent();
        }
    }

    protected void calcMinMaxY(TreeValue e) {
        if (e == null || !e.isEnable()) return;

        float high = e.getHigh();
        float low = e.getLow();
        if (!Float.isNaN(high) && !Float.isInfinite(high)) {
            mViewportYMin = Math.min(mViewportYMin, low);
            mViewportYMax = Math.max(mViewportYMax, high);
        }
    }

    @Override
    public void setValues(List<TreeValue> values) {
        this.mTreeValues = values;
    }

    @Override
    public List<TreeValue> getValues() {
        return mTreeValues;
    }

    @Override
    public boolean addEntry(TreeValue e) {
        if (e == null) return false;

        if (mTreeValues == null) {
            mTreeValues = new ArrayList<>();
        }

        calcMinMaxY(e);

        return mTreeValues.add(e);
    }

    @Override
    public boolean removeEntry(TreeValue e) {
        if (e == null) return false;
        calcMinMaxY(e);
        return mTreeValues.remove(e);
    }

    @Override
    public int getEntryIndex(TreeValue e) {
        return mTreeValues.indexOf(e);
    }

    @Override
    public TreeValue getEntryForIndex(int index) {
        return mTreeValues.get(index);
    }

    public float getStrokeThickness() {
        return strokeThickness;
    }

    public void setStrokeThickness(float strokeThickness) {
        this.strokeThickness = strokeThickness;
    }

    public int getPositiveColor() {
        return positiveColor;
    }

    public void setPositiveColor(int positiveColor) {
        this.positiveColor = positiveColor;
    }

    public int getNegativeColor() {
        return negativeColor;
    }

    public void setNegativeColor(int negativeColor) {
        this.negativeColor = negativeColor;
    }

    public int getColorAlpha() {
        return colorAlpha;
    }

    public void setColorAlpha(int colorAlpha) {
        if (colorAlpha < 0) {
            this.colorAlpha = 0;
            return;
        }
        if (colorAlpha > 255) {
            this.colorAlpha = 255;
            return;
        }
        this.colorAlpha = colorAlpha;
    }
}

File: chart/src/main/java/cn/jingzhuan/lib/chart3/data/dataset/TreeDataSet.kt
package cn.jingzhuan.lib.chart3.data.dataset

import android.graphics.Color
import cn.jingzhuan.lib.chart3.Viewport
import cn.jingzhuan.lib.chart3.axis.AxisY
import cn.jingzhuan.lib.chart3.axis.AxisY.AxisDependency
import cn.jingzhuan.lib.chart3.data.value.TreeValue
import java.lang.Float.isInfinite
import java.lang.Float.isNaN
import kotlin.math.max
import kotlin.math.min

/**
 * @since 2023-09-05
 * created by lei
 */
open class TreeDataSet : AbstractDataSet<TreeValue> {

    constructor(treeValues: List<TreeValue>) : this(treeValues, AxisY.DEPENDENCY_BOTH)

    constructor(treeValues: List<TreeValue>, @AxisDependency axisDependency: Int) : super(treeValues, axisDependency)

    var strokeThickness = 2f

    var positiveColor = Color.RED

    var negativeColor = Color.GREEN

    var colorAlpha = 255
        set(colorAlpha) {
            if (colorAlpha < 0) {
                field = 0
                return
            }
            if (colorAlpha > 255) {
                field = 255
                return
            }
            field = colorAlpha
        }

    override fun getEntryCount(): Int {
        return values.size
    }

    override fun calcMinMax(viewport: Viewport) {
        if (values.isEmpty()) return
        viewportYMax = -Float.MAX_VALUE
        viewportYMin = Float.MAX_VALUE

        val visiblePoints = getVisiblePoints(viewport)

        if (visiblePoints.isNullOrEmpty()) return

        for (value in visiblePoints) {
            calcMinMaxY(value)
        }

        val range = viewportYMax - viewportYMin
        if (minValueOffsetPercent.compareTo(0f) > 0f) {
            viewportYMin -= range * minValueOffsetPercent
        }
        if (maxValueOffsetPercent.compareTo(0f) > 0f) {
            viewportYMax += range * maxValueOffsetPercent
        }
    }

    private fun calcMinMaxY(value: TreeValue?) {
        if (value == null || !value.isEnable) return
        val high = value.high
        val low = value.low
        if (!isNaN(high) && !isInfinite(high)) {
            viewportYMin = min(viewportYMin, low)
            viewportYMax = max(viewportYMax, high)
        }
    }

    override fun addEntry(value: TreeValue?): Boolean {
        if (value == null) return false
        calcMinMaxY(value)
        return values.add(value)
    }

    override fun removeEntry(value: TreeValue?): Boolean {
        if (value == null) return false
        calcMinMaxY(value)
        return values.remove(value)
    }
}
